### PR TITLE
Support output dot

### PR DESCRIPTION
**Pull request by ParkHanbum**
committed to support output format of dot.
associated with issue #355

### CODE REVIEW

File: cmds/dump.c [hunk 1]
```diff
@@ -69,6 +69,12 @@ struct uftrace_raw_dump {
 	uint64_t kbuf_offset;
 };
 
+struct uftrace_dot_dump {
+	struct uftrace_dump_ops ops;
+	uint64_t file_offset;
+	uint64_t kbuf_offset;
```
Comment (namhyung) [inline cmds/dump.c]: Where are these two fields used?

File: cmds/dump.c [hunk 2]
```diff
@@ -1209,6 +1215,130 @@ static void print_flame_footer(struct uftrace_dump_ops *ops,
...
+			     struct ftrace_file_handle *handle,
+			     struct opts *opts)
+{
+	pr_out("# version\":\"uftrace %s\",\n", UFTRACE_VERSION);
+
+	if (handle->hdr.info_mask & (1UL << CMDLINE))
+		pr_out("# command_line \"%s\"\n", handle->info.cmdline);
+
+	pr_out("digraph \"");
+	print_flame_header(ops, handle, opts);
```
Comment (namhyung) [inline cmds/dump.c]: Why do you use `print_flame_header()` here?

File: cmds/dump.c [hunk 3]
```diff
@@ -1209,6 +1215,130 @@ static void print_flame_footer(struct uftrace_dump_ops *ops,
...
+
+static void print_dot_inverted_time(struct uftrace_dump_ops *ops,
+				    struct ftrace_task_handle *task)
+{
+}
+
+static void print_dot_task_rstack(struct uftrace_dump_ops *ops,
+				  struct ftrace_task_handle *task, char *name)
+{
+	print_flame_task_rstack(ops, task, name);
```
Comment (namhyung) [inline cmds/dump.c]: Please do not use flame graph data structure for other purpose.  Or make it to safe for sharing (I guess it's already so, but..)

File: cmds/dump.c [hunk 4]
```diff
@@ -1209,6 +1215,130 @@ static void print_flame_footer(struct uftrace_dump_ops *ops,
...
+static void print_dot_perf_start(struct uftrace_dump_ops *ops,
+				   struct uftrace_perf_reader *perf, int cpu)
+{
+}
+
+static void print_dot_perf_event(struct uftrace_dump_ops *ops,
+				   struct uftrace_perf_reader *perf,
+				   struct uftrace_record *frs)
+{
+}
```
Comment (namhyung) [inline cmds/dump.c]: Maybe we should remove this empty callbacks and check it before calling.

File: cmds/dump.c [hunk 5]
```diff
@@ -1209,6 +1215,130 @@ static void print_flame_footer(struct uftrace_dump_ops *ops,
...
+
+static void print_fg_to_dot(struct uftrace_graph_node *node,
+				struct opts *opts)
+{
+	struct uftrace_graph_node *child;
+	unsigned long sample = node->nr_calls;
+
+	if (sample && opts->sample_time)
+		sample = (node->time - node->child_time) / opts->sample_time;
```
Comment (namhyung) [inline cmds/dump.c]: This sampling logic is really needed?

File: cmds/dump.c [hunk 6]
```diff
@@ -1209,6 +1215,130 @@ static void print_flame_footer(struct uftrace_dump_ops *ops,
...
+			pr_out("%s", parent->name);
+			pr_out("->");
+		}
+		pr_out("%s", node->name);
+
+		// Edge Attributes
+		pr_out("[");
+		pr_out("xlabel = \"Calls : %lu\"", sample);
+		pr_out("]");
+		pr_out("\n");
```
Comment (namhyung) [inline cmds/dump.c]: Hmm.. why not making it a single line like `pr_out("[xlabel = \"Calls: %lu\"]\n", sample);`?

File: uftrace.c [hunk 7]
```diff
@@ -153,6 +154,7 @@ static struct argp_option uftrace_options[] = {
 	{ "kernel-only", OPT_kernel_only, 0, 0, "Dump kernel data only" },
 	{ "flame-graph", OPT_flame_graph, 0, 0, "Dump recorded data in FlameGraph format" },
 	{ "sample-time", OPT_sample_time, "TIME", 0, "Show flame graph with this sampling time" },
+	{ "dot", OPT_dot, 0, 0, "Dump recorded data in DOT format" },
```
Comment (namhyung) [inline uftrace.c]: I think the name "dot" is too short and vague.  I prefer "graphviz" instead - what do you think?

File: cmds/dump.c [hunk 8]
```diff
@@ -1491,67 +1385,63 @@ int command_dump(int argc, char *argv[], struct opts *opts)
...
+		.task_event     = NULL,
+		.kernel_start   = NULL,
+		.cpu_start      = NULL,
+		.kernel_func    = NULL,
+		.kernel_event   = NULL,
+		.lost           = NULL,
+		.perf_start     = NULL,
+		.perf_event     = NULL,
+		.footer         = NULL,
+	};
```
Comment (namhyung) [inline cmds/dump.c]: No need to have this struct separately IMO.

File: cmds/dump.c [hunk 9]
```diff
@@ -1491,67 +1385,63 @@ int command_dump(int argc, char *argv[], struct opts *opts)
...
-				.kernel_start   = print_chrome_kernel_start,
-				.cpu_start      = print_chrome_cpu_start,
-				.kernel_func    = print_chrome_kernel_rstack,
-				.kernel_event   = print_chrome_kernel_event,
-				.lost           = print_chrome_kernel_lost,
-				.perf_start     = print_chrome_perf_start,
-				.perf_event     = print_chrome_perf_event,
-				.footer         = print_chrome_footer,
-			},
+			.ops = ops,
```
Comment (namhyung) [inline cmds/dump.c]: Please set the ops here as before.

File: utils/utils.h [hunk 10]
```diff
@@ -209,6 +209,12 @@ extern void setup_signal(void);
 	}								\
 })
 
+#define call_if_nonull(fptr, ...) 					\
+({									\
+	if ( fptr != NULL )						\
```
Comment (namhyung) [inline utils/utils.h]: I don't like the extra whitespaces.

File: uftrace.c [hunk 11]
```diff
@@ -153,6 +154,7 @@ static struct argp_option uftrace_options[] = {
 	{ "kernel-only", OPT_kernel_only, 0, 0, "Dump kernel data only" },
 	{ "flame-graph", OPT_flame_graph, 0, 0, "Dump recorded data in FlameGraph format" },
 	{ "sample-time", OPT_sample_time, "TIME", 0, "Show flame graph with this sampling time" },
+	{ "graphviz", OPT_graphviz, 0, 0, "Dump recorded data in DOT format" },
```
Comment (namhyung) [inline uftrace.c]: You also need to update the documentation.

File: cmds/dump.c [hunk 12]
```diff
@@ -1103,6 +1107,90 @@ static void print_flame_footer(struct uftrace_dump_ops *ops,
...
+	struct uftrace_graph_node *child;
+	unsigned long n_calls = node->nr_calls;
+
+	if (n_calls) {
+		struct uftrace_graph_node *parent = node->parent;
+
+		pr_out("\t");
+		if (parent != NULL && parent->name != NULL) {
+			pr_out("%s", parent->name);
+			pr_out("->");
```
Comment (namhyung) [inline cmds/dump.c]: Can be merged to a single line.

File: uftrace.c (hunk 11 above)
Comment (honggyukim) [reply uftrace.c]: IMHO, `--dot` looks simpler than `--graphviz`.

File: utils/utils.h [hunk 13]
```diff
@@ -211,7 +211,7 @@ extern void setup_signal(void);
 
 #define call_if_nonull(fptr, ...) 					\
 ({									\
-	if ( fptr != NULL )						\
+	if (fptr != NULL)						\
```
Comment (namhyung) [inline utils/utils.h]: Oh, please fold this into the first commit.

File: cmds/dump.c [hunk 14]
```diff
@@ -1495,16 +1389,16 @@ int command_dump(int argc, char *argv[], struct opts *opts)
...
-				.kernel_event   = print_chrome_kernel_event,
-				.lost           = print_chrome_kernel_lost,
-				.perf_start     = print_chrome_perf_start,
+				.task_event     = NULL,
+				.kernel_start   = NULL,
+				.cpu_start      = NULL,
+				.kernel_func    = NULL,
+				.kernel_event   = NULL,
+				.lost           = NULL,
+				.perf_start     = NULL,
```
[the remark this answers is not in the document]
Comment (ParkHanbum) [reply cmds/dump.c]: cool!

File: cmds/dump.c [hunk 15]
```diff
@@ -82,6 +82,10 @@ struct uftrace_flame_dump {
 	uint64_t sample_time;
 };
 
+struct uftrace_graphviz_dump {
+	struct uftrace_dump_ops ops;
+};
+
```
Comment (namhyung) [inline cmds/dump.c]: Please move the graphviz changes to the next commit

**Comment by namhyung** [comment]
Please update!

**Comment by ParkHanbum** [comment]
@namhyung updated!

**Comment by namhyung** [comment]
Please don't put the graphviz changes in the first commit - it should only have `call_if_nonull()` changes, ok?

Also it'd be really nice if you come up with a test case for this.

**Comment by ParkHanbum** [comment]
@namhyung I have updated this issue with testcase.

**Comment by ParkHanbum** [comment]
@namhyung it's my pleasure~

File: cmds/dump.c [hunk 16]
```diff
@@ -82,6 +82,10 @@ struct uftrace_flame_dump {
 	uint64_t sample_time;
 };
 
+struct uftrace_graphviz_dump {
+	        struct uftrace_dump_ops ops;
```
Comment (namhyung) [inline cmds/dump.c]: wrong indentation

File: cmds/dump.c [hunk 17]
```diff
@@ -1111,6 +1115,89 @@ static void print_flame_footer(struct uftrace_dump_ops *ops,
...
+				       char *name)
+{
+	struct uftrace_record *frs = task->rstack;
+	struct uftrace_task_graph *graph;
+
+	graph = graph_get_task(task, sizeof(*graph));
+
+	graph->graph = &graphviz_graph;
+	graphviz_graph.sess = find_task_session(&task->h->sessions,
+			task->tid, frs->time);
```
Comment (namhyung) [inline cmds/dump.c]: one more indentation nit: the argument line should start from the same position with the above line.

File: tests/t207_dump_graphviz.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
+#!/usr/bin/env python
+
+from runtest import TestBase
+import subprocess as sp
+
+TDIR='xxx'
+
+class TestCase(TestBase):
+    def __init__(self):
+        TestBase.__init__(self, 'abc', """digraph "/home/m/git/uftrace/tests/t-abc" {
```
Comment (namhyung) [inline tests/t207_dump_graphviz.py]: Will this line succeed on other machines?

Comment (namhyung) [reply tests/t207_dump_graphviz.py]: Please don't say that as we (including me) always do mistakes.. :)

**Pull request merged**